p2p/router.go

```diff
@@ -0,0 +1,568 @@
+		}()
+	}
+}
+
+// dialPeers maintains outbound connections to peers.
+func (r *Router) dialPeers() {
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		default:
+		}
+
+		peer := r.store.Dispense()
+		if peer == nil {
+			r.logger.Debug("no eligible peers, sleeping")
+			select {
+			case <-time.After(time.Second):
+				continue
+			case <-r.stopCh:
+				return
+			}
+		}
+
+		go func() {
+			defer r.store.Return(peer.ID)
+			conn, err := r.dialPeer(peer)
+			if err != nil {
+				r.logger.Error("failed to dial peer, will retry", "peer", peer.ID)
+				return
+			}
+			defer conn.Close()
+
+			queue := newFIFOQueue()
+			defer queue.close()
+			r.peerMtx.Lock()
+			r.peerQueues[peer.ID] = queue
+			r.peerMtx.Unlock()
+
+			defer func() {
+				r.peerMtx.Lock()
+				delete(r.peerQueues, peer.ID)
+				r.peerMtx.Unlock()
+			}()
+
+			r.routePeer(peer.ID, conn, queue)
+		}()
+	}
+}
+
+// dialPeer attempts to connect to a peer.
+func (r *Router) dialPeer(peer *peerInfo) (Connection, error) {
+	ctx := context.Background()
+
+	for _, address := range peer.Addresses {
+		resolveCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		r.logger.Info("resolving peer address", "peer", peer.ID, "address", address)
+		endpoints, err := address.Resolve(resolveCtx)
+		if err != nil {
+			r.logger.Error("failed to resolve address", "address", address, "err", err)
+			continue
+		}
+
+		for _, endpoint := range endpoints {
+			t, ok := r.transports[endpoint.Protocol]
+			if !ok {
+				r.logger.Error("no transport found for protocol", "protocol", endpoint.Protocol)
+				continue
+			}
+			dialCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+			conn, err := t.Dial(dialCtx, endpoint)
+			if err != nil {
+				r.logger.Error("failed to dial endpoint", "endpoint", endpoint)
+			} else {
+				r.logger.Info("connected to peer", "peer", peer.ID, "endpoint", endpoint)
+				return conn, nil
+			}
+		}
+	}
+	return nil, errors.New("failed to connect to peer")
+}
+
+// routePeer routes inbound messages from a peer to channels, and also sends
+// outbound queued messages to the peer. It will close the connection and send
+// queue, using this as a signal to coordinate the internal receivePeer() and
+// sendPeer() goroutines. It blocks until the peer is done, e.g. when the
+// connection or queue is closed.
+func (r *Router) routePeer(peerID NodeID, conn Connection, sendQueue queue) {
+	// FIXME: Peer updates should probably be handled by the peer store.
+	r.peerUpdatesCh <- PeerUpdate{
+		PeerID: peerID,
+		Status: PeerStatusUp,
+	}
+	defer func() {
+		r.peerUpdatesCh <- PeerUpdate{
+			PeerID: peerID,
+			Status: PeerStatusDown,
+		}
+	}()
+
+	resultsCh := make(chan error, 2)
+	go func() {
+		resultsCh <- r.receivePeer(peerID, conn)
+	}()
+	go func() {
+		resultsCh <- r.sendPeer(peerID, conn, sendQueue)
+	}()
+
+	err := <-resultsCh
+	_ = conn.Close()
+	sendQueue.close()
+	if e := <-resultsCh; err == nil {
```

Comment:
golang.org/x/sync/errgroup ?

Reply:
Nice, wasn't aware of `errgroup`. However, in this case we're using `conn.Close()` and `sendQueue.close()` to signal termination of the other goroutines after one of them returns, unlike `errgroup` which relies to `context` for signaling. I think introducing contexts in addition to connections and queues here just complicates things.